src/emulator/gdbstub/src/gdb.cpp

```diff
@@ -0,0 +1,597 @@
+// Vita3K emulator project
+// Copyright (C) 2018 Vita3K team
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+#ifdef USE_GDBSTUB
+
+#include <host/state.h>
+#include <util/log.h>
+
+#include <gdbstub/functions.h>
+
+#include <cpu/functions.h>
+#include <kernel/functions.h>
+
+#include <spdlog/fmt/bundled/printf.h>
+#include <sstream>
+
+// Sockets
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define LOG_GDB_EVERYTHING
+//#define LOG_GDB_DEBUG
+
+#ifdef LOG_GDB_EVERYTHING
+#define LOG_GDB_DEBUG
+#endif
+
+#ifdef LOG_GDB_DEBUG
+#define LOG_GDB LOG_INFO
+#else
+#define LOG_GDB(a, ...)
+#endif
+
+// Based off RPCS3's implementation.
+
+typedef char PacketData[1200];
+
+struct PacketCommand {
+    char *data{};
+    ssize_t length = -1;
+
+    ssize_t begin_index = -1;
+    ssize_t end_index = -1;
+    ssize_t content_length = -1;
+
+    char *content_start{};
+    char *checksum_start{};
+
+    uint8_t checksum = 0;
+
+    bool is_valid = false;
+};
+
+typedef std::function<std::string(HostState &state, PacketCommand &command)> PacketFunction;
+
+struct PacketFunctionBundle {
+    std::string name;
+    PacketFunction function;
+};
+
+// DEBUG + LOGGING
+std::string pd_string(const char *data, ssize_t length) {
+    return std::string(data, static_cast<unsigned long>(length));
+}
+
+std::string content_string(PacketCommand &command) {
+    return std::string(command.content_start, static_cast<unsigned long>(command.content_length));
+}
+
+// Hexes
+std::string be_hex(uint32_t value) {
+    return fmt::format("{:0>8x}", htonl(value));
+}
+
+std::string to_hex(uint32_t value) {
```

Comment:
Use [`this`](https://github.com/Vita3K/Vita3K/blob/master/src/emulator/util/include/util/log.h#L59) instead (add the `>` thing in there, too).
Because it's generic and there's no `8` it might not work. Test it, and if it doesn't remove the template and make multiple versions for u32, u64 and whatever else we use.
Or if you feel fancy you can probably still make that compile time with templates, but it's not trivial and there's no real need.

Reply:
I would, but I would have to remove the `0x` in that method too (gdb already knows its expecting raw hexes). Maybe I could add another method? One that spits out raw hexes and another that adds the `0x` at the beginning?

Reply:
Ah ok. No worries then, use your own function.